File: calculateExecutionTimes/calculateExecutionTimes.js
'use strict';

const assert = require('node:assert');

class FunctionMetadata {
    name;
    #startTime = undefined;
    #endTime = undefined;

    constructor(name) {
        this.name = name;
    }

    addEvent(type, time) {
        assert.match(type, /start|end/, 'Type value not one of "start" or "end"');

        if (type === 'start') {
            this.#startTime = time;
        }
        else if (type === 'end') {
            this.#endTime = time;
        }
    }

    getExecutionTime() {
        assert.notEqual(this.#startTime, undefined, 'No start event found');
        assert.notEqual(this.#endTime, undefined, 'No end event found');

        return this.#endTime - this.#startTime;
    }
}

const functionMetadataMap = new Map();

function processFunctionEvents(functionEvents) {
    functionEvents.forEach((functionEvent) => {
        const { name, time, event: type } = functionEvent;
        let functionMetadata;

        if (functionMetadataMap.has(name)) {
            functionMetadata = functionMetadataMap.get(name);
        }
        else {
            functionMetadata = new FunctionMetadata(name);
            functionMetadataMap.set(name, functionMetadata);
        }

        functionMetadata.addEvent(type, time);
    });
}

function calculateExecutionTimes(functionEvents) {
    processFunctionEvents(functionEvents);

    const result = {};

    functionMetadataMap.forEach((functionMetadata) => {
        result[functionMetadata.name] = functionMetadata.getExecutionTime();
    });

    return result;
}

const tests = [
    [
        { name: 'main', time: 0, event: 'start' },
        { name: 'subTask1', time: 5, event: 'start' },
        { name: 'subTask1', time: 10, event: 'end' },
        { name: 'subTask2', time: 15, event: 'start' },
        { name: 'main', time: 25, event: 'end' }
    ],
    [
        { name: 'main', time: 0, event: 'start' },
        { name: 'subTask1', time: 5, event: 'start' },
        { name: 'subTask1', time: 10, event: 'stop' },
        { name: 'subTask2', time: 15, event: 'start' },
        { name: 'main', time: 25, event: 'end' }
    ],
    [
        { name: 'main', time: 0, event: 'start' },
        { name: 'subTask1', time: 5, event: 'start' },
        { name: 'subTask1', time: 10, event: 'end' },
        { name: 'subTask2', time: 15, event: 'start' },
        { name: 'subTask2', time: 20, event: 'end' },
        { name: 'main', time: 25, event: 'end' }
    ],
    [
        { name: 'subTask2', time: 15, event: 'start' },
        { name: 'subTask1', time: 10, event: 'end' },
        { name: 'subTask2', time: 20, event: 'end' },
        { name: 'subTask1', time: 5, event: 'start' },
        { name: 'main', time: 25, event: 'end' },
        { name: 'main', time: 0, event: 'start' }
    ]
];

tests.forEach((test) => {
    console.log(`calculateExecutionTimes(${JSON.stringify(test)});`);

    try {
        console.log(calculateExecutionTimes(test));
    }
    catch (e) {
        console.error(e.message);
    }
});

File: calculateExecutionTimes/calculateExecutionTimes.py
class FunctionMetadata:
    name = None
    __startTime = None
    __endTime = None

    def __init__(self, name):
        self.name = name

    def addEvent(self, type, time):
        assert type in ['start', 'end'], 'Type value not one of "start" or "end"'

        if (type == 'start'):
            self.__startTime = time
        elif (type == 'end'):
            self.__endTime = time
    
    def getExecutionTime(self):
        assert not (self.__startTime is None), 'No start event found'
        assert not (self.__endTime is None), 'No end event found'

        return self.__endTime - self.__startTime

functionMetadataMap = {}

def processFunctionEvents(functionEvents):
    for functionEvent in functionEvents:
        name = functionEvent['name']
        functionMetadata = None

        if (name in functionMetadataMap):
            functionMetadata = functionMetadataMap[name]
        else:
            functionMetadata = FunctionMetadata(name)
            functionMetadataMap[name] = functionMetadata

        functionMetadata.addEvent(functionEvent['event'], functionEvent['time'])

def calculateExecutionTimes(functionEvents):
    processFunctionEvents(functionEvents)

    result = {}

    for functionMetadata in functionMetadataMap.values():
        result[functionMetadata.name] = functionMetadata.getExecutionTime()
    
    return result

tests = [
    [
        { 'name': 'main', 'time': 0, 'event': 'start' },
        { 'name': 'subTask1', 'time': 5, 'event': 'start' },
        { 'name': 'subTask1', 'time': 10, 'event': 'end' },
        { 'name': 'subTask2', 'time': 15, 'event': 'start' },
        { 'name': 'main', 'time': 25, 'event': 'end' }
    ],
    [
        { 'name': 'main', 'time': 0, 'event': 'start' },
        { 'name': 'subTask1', 'time': 5, 'event': 'start' },
        { 'name': 'subTask1', 'time': 10, 'event': 'stop' },
        { 'name': 'subTask2', 'time': 15, 'event': 'start' },
        { 'name': 'main', 'time': 25, 'event': 'end' }
    ],
    [
        { 'name': 'main', 'time': 0, 'event': 'start' },
        { 'name': 'subTask1', 'time': 5, 'event': 'start' },
        { 'name': 'subTask1', 'time': 10, 'event': 'end' },
        { 'name': 'subTask2', 'time': 15, 'event': 'start' },
        { 'name': 'subTask2', 'time': 20, 'event': 'end' },
        { 'name': 'main', 'time': 25, 'event': 'end' }
    ],
    [
        { 'name': 'subTask2', 'time': 15, 'event': 'start' },
        { 'name': 'subTask1', 'time': 10, 'event': 'end' },
        { 'name': 'subTask2', 'time': 20, 'event': 'end' },
        { 'name': 'subTask1', 'time': 5, 'event': 'start' },
        { 'name': 'main', 'time': 25, 'event': 'end' },
        { 'name': 'main', 'time': 0, 'event': 'start' }
    ]
]

for test in tests:
    print(f'calculateExecutionTimes(${test})')

    try:
        print(calculateExecutionTimes(test))
    except AssertionError as e:
        print(str(e))
